### DISCUSSION TITLE
How to avoid creating multiple sessions when using Dependencies with Security?

### DISCUSSION THREAD

**Discussion by DurandA** [fastapi/fastapi#6024] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code ([full working example](https://github.com/DurandA/fastapi-security-example))

```python
from fastapi import Depends, Security
from fastapi.security import SecurityScopes
from sqlmodel import Session, create_engine, select


def get_session():
    with Session(engine) as session:
        yield session


async def get_current_user(
    security_scopes: SecurityScopes,
    token: str = Depends(oauth2_scheme),
    session: Session = Depends(get_session)
):
    # Token read omitted 👈
    user = session.get(User, token_data.user_uuid)
    if user is None:
        raise credentials_exception
    return user


async def get_current_app_user(
    current_user: User = Security(get_current_user, scopes=["app"])
):
    return current_user


@app.patch("/me", response_model=UserRead)
def update_current_user(
    user: UserUpdate,
    current_user: User = Depends(get_current_app_user),
    session: Session = Depends(get_session),
):
    user_data = user.dict(exclude_unset=True)
    for key, value in user_data.items():
        setattr(current_user, key, value)
    # Session and session from current_user are different ⚠️
    session.add(current_user) 
    session.commit()
    session.refresh(current_user)
    return current_user
```

### Description

I am using a synchronous (SQLite) session in a FastAPI Dependency as shown in the [SQLModel Tutorial](https://sqlmodel.tiangolo.com/tutorial/fastapi/session-with-dependency/#create-a-fastapi-dependency).

When a route "depends" on a `Dependency` and a `Security` that use `get_session`, the session is created twice resulting in errors such as:
```
sqlalchemy.exc.InvalidRequestError: Object '<User at 0x7f218cbf6a00>' is already attached to session '1' (this is '2')
```

How can this be prevented? I would expect a Security to [reuse Dependencies](https://fastapi.tiangolo.com/tutorial/dependencies/sub-dependencies/#using-the-same-dependency-multiple-times) as long as `use_cache=True`.

### Operating System

Linux

### Operating System Details

Ubuntu 20.04

### FastAPI Version

0.89.1

### Python Version

Python 3.8.10

### Additional Context

The question was originally asked in [SQLModel discussion](https://github.com/tiangolo/sqlmodel/discussions/555). However, I think this is unrelated to SQLModel and maybe bug with FastAPI `Security` object. I also created a [StackOverflow question with a bounty](https://stackoverflow.com/questions/75487025/how-to-avoid-creating-multiple-sessions-when-using-fastapi-dependencies-with-sec) if you have a nice solution on how to handle this.

**Comment by yinziyan1206** [comment]
The `cache_key` of `Security` is different with that of `Depends`. it has to be two sessions if you use di. You can return the session in `get_current_user` or create session in middleware.

**Reply by YuriiMotov** [reply]
I think none of these options is a good solution.
`get_current_user` should return `user`, not `session`. And when you create session in middleware you can't override it easily.

Could you take a look at [this post](https://github.com/tiangolo/fastapi/discussions/6024#discussioncomment-8541913) and give your opinion?

**Comment by DurandA** [comment]
This issue appears only if `scopes` of `Security` is a non empty list. This is because [`scopes` are part of the `cache_key`](https://github.com/tiangolo/fastapi/blob/7b3727e03e84ca202d450ba3d702d5cd37025d60/fastapi/dependencies/models.py#L58):
```python
self.cache_key = (self.call, tuple(sorted(set(self.security_scopes or []))))
```

I believe the intended behavior is to isolate `Depends` with different `securit_scopes`. I am however not sure why it should change the scope of parent Dependencies and not only their children. Also it would be nice to mark a Dependency with something like `ignore_scopes` parameter.

**Comment by DurandA** [comment]
See https://github.com/tiangolo/fastapi/pull/2945 about why the isolation of cache key with scopes was introduced.

**Comment by YuriiMotov** [comment]
I would like to resume the discussion on dependency caching.

I apologize in advance for the long text..

The documentation says that

> If one of your dependencies is declared multiple times for the same path operation, for example, multiple dependencies have a common sub-dependency, FastAPI will know to call that sub-dependency only once per request.

But this is not true. In the current implementation, FastAPI caches dependencies according to the dependency name and scopes. Thus, one dependency that is used twice with different scopes will be called twice (and cached twice).

This is not entirely intuitive behavior and it is undesirable in some cases.
For example, this leads to the opening of two database sessions in the following example:
<img width="774" alt="1" src="https://github.com/tiangolo/fastapi/assets/109919500/2098380d-1e70-4d03-b57b-50eda32205e3">

In the example above, `get_db_session` will be called twice because scopes are different ([] and ["me"]).

There are a few discussions on this topic:
1. This one
2. Discussion in the comments of PR https://github.com/tiangolo/fastapi/pull/2945
2. https://github.com/tiangolo/fastapi/discussions/9513

And @[DurandA](https://github.com/DurandA) has created a PR to fix this problem https://github.com/tiangolo/fastapi/pull/9790.
That PR fixes that particular problem, and `get_db_session` will be called only once:
<img width="489" alt="2" src="https://github.com/tiangolo/fastapi/assets/109919500/a59cf9f3-acbe-45bc-aa4e-53bb062ff3a8">

Even though this PR doesn't brake any other tests, I think it brakes the logic of scopes propagation.
Let's take a look at the following example:
<img width="759" alt="3" src="https://github.com/tiangolo/fastapi/assets/109919500/377d9981-1790-482a-a6ff-d6c41195247d">

Let's assume that `get_current_user` uses `security_scopes`.
In current implementation (FastAPI 0.109.2) FastAPI will call `get_current_user` twice with different scopes ([] and ["me"]).
And after merging of this PR (https://github.com/tiangolo/fastapi/pull/9790) FastAPI will call it only once (with scopes = []).
I think that's incorrect.

Before summarizing, let`s take a look at one more example:
<img width="1046" alt="4" src="https://github.com/tiangolo/fastapi/assets/109919500/092d6458-ff7f-480a-bc56-13a0c963d21c">

Let`s assume that:

- `get_db_session` doesn't use security_scopes in any way and it's subdependencies also doesn`t use it
- `get_current_user` uses security_scopes
- `get_something` doesn't use security_scopes itself, but it uses subdependency `get_current_user` that uses security_scopes
- `get_something_2` doesn't use security_scopes itself, but it uses subdependency `get_something` that uses subdependency `get_current_user` that uses security_scopes

From my point of view it would be logical to call `get_db_session` once and cache it using key={function_name} (without scopes).
But both of `get_current_user` and `get_something` must be called twice and cached with the keys that include scopes.
Do you agree? ** Is this how it was supposed to work?**

I think it's important to take a deeper look and figure it out before making changes to avoid new mistakes.
I suggest discussing this issue and implementing this logic as it was intended. And we need to fix the documentation (describe dependency caching rules) along the way.

**Reply by DurandA** [reply]
Thank you for taking the time to document the issue with my PR.

**Reply by DurandA** [reply]
Do you have a particular implementation is mind to fix the documented issue?

**Reply by YuriiMotov** [reply]
No, not yet.
I think it's better to discuss how it should work before proceeding with the implementation.
Let's try to get more people involved in this discussion

**Reply by rishabhc32** [reply]
@YuriiMotov I think first we should include this caveat in the documentation so users are aware of it in advance.

Also I agree with @DurandA that adding an optional `ignore_scopes` parameter can help, where scopes are propagated but not used for cache keys.

**Reply by rishabhc32** [reply]
Does @tiangolo have any views on it?

**Reply by thibautd** [reply]
Thank you for summarizing all this, I've spent hours trying to figure out what is happening. That behavior doesn't match the documentation, especially regarding `SecurityScopes` which never includes all the scopes in the path, but instead the security dependency is called multiple time for each scope. We won't use the "scopes" feature of FastAPI as it looks unreliable at moment, and fallback to classic dependencies per expected scopes. Thanks again for this very nice summary.

**Reply by qmorek** [reply]
@YuriiMotov / @tiangolo is anything happening with that topic?

**Reply by tiangolo** [reply]
Thanks a lot for the research and diagrams @YuriiMotov! This was very useful. In fact, I made one of the tests precisely reflecting the last diagram ideas. :rocket:

**Comment by tiangolo** [comment, marked as answer]
Thanks for the discussion everyone! :cake: 

This should be solved by https://github.com/fastapi/fastapi/pull/14419, released in FastAPI 0.123.0 :tada:
